packages/opentelemetry-plugin-http/src/http.ts

```diff
@@ -200,6 +201,7 @@ export class HttpPlugin extends BasePlugin<Http> {
           { hostname }
         );
         span.setAttributes(attributes);
+        this._onNewResponse(span, response);
```

Comment:
I think that probably it would be better if you modify this whole concept a bit because of performance issues.

First Approach
During a plugin initialisation check if user defined a hooks in config if that is the case then add simply a different event listener (request.on) with your modifications `plugin._onNewRequest` etc. otherwise use the one without your modifications.
This will have some code duplication but in fact 0 performance issue if user doesn't define any hooks.

Second Approach 
Also during initialisation check if user defined a hooks in config but this time add separate event listener and then call `plugin._onNewRequest`.
This could be achieved by adding extra parameter to `_traceClientRequest`.
This would be basically a callback for setting up the desired events, but if user doesn't define the hooks this callback will be simply something like  `NoopHookCallback` which will look like `function () {}` and can be called before returning the request inside function `_traceClientRequest`

WDYT ?

Reply:
Thank you for the review,

Regarding the performance impact, when these function are not set, the only performance issue is function call to `_onNewRequest` \ `_onNewResponse`, and executing an if: `if (this._config.requestHook)`. Is it something to worry about?

On plugin initialisation I still have no `request`\`response` objects on which I can add events. I call the function to register events as soon as the `request` object is available (in incoming or outgoing functions).

What I can do is to register on the `http.Server` `request` Event, but then I'll get a callback without the relevant span, and it will be called every-time, also when the request is ignored in open telemetry.

Reply:
>Regarding the performance impact, when these function are not set, the only performance issue is function call to _onNewRequest \ _onNewResponse, and executing an if: if (this._config.requestHook). Is it something to worry about?

Yes. When this is done on every single request and response on both client and server it is something to worry about. Function call overhead is non-negligible. Right now you have a function which is called every time and an if-check that happens inside the function. Moving the if-check outside of the function and only calling it if there is something configured would be drastically more performant.

Reply:
Thanks @dyladan 
I wasn't aware of the impact of the function call. Will perform the if-check before calling the function and push a new commit

Reply:
I think the change you did will have exactly the same number of checks then previously

```
request.on('some event', a);

function a() {
  b();
  console.log('bar');
}

function b() {
  if (config.c) {
    console.log('foo');
  }
}
```

is exactly the same as

```
request.on('some event', a);

function a() {
  if (config.c) {
    b();
  }
  console.log('bar');
}

function b() {
  console.log('foo');
}
```

What I tried to say is to refactor into something like this

```
if (config.c) {
  request.on('some event', a1);
} else {
  request.on('some event', a2);
}

function a1() {
  b();
  console.log('bar');
}

function a2() {
  console.log('bar');
}

function b() {
  console.log('foo');
}

```

this way you will have some duplication in code, but when it is about performance, you will check only once for `config.c` rather then every time on `some event`

Reply:
Sorry for the delay, missed the comment.
I like the idea of registering to the `EventEmitter` interface.
In this case, the hook is exposing the `request` object (which we get as the patched function parameter) so the user is able to do function calls like `request.on('some event', a1);`

Reply:
To be more precise, the above code still execute the `if (config.c)` **per request** which is the same as what current implementation is doing